### DISCUSSION TITLE
Solidus rails payment gateway integration with callback url

### DISCUSSION THREAD

**Discussion by raunjon** [solidusio/solidus#3915] [Troubleshooting]
I am integrating a payment gateway with my Solidus app, the payment gateway I am using is 'Razorpay'

The payment gateway, handles the payment through a modal popup on their servers directly, and then initiates a callback url on the success or failure of the payment.

How do I handle this in Solidus? I created a POST request to capture the details, at 'checkout/verify_payment', where I verify the payment and check if its a success or a failure, I am not sure how do I proceed on the success callback.

```
  def verify_payment
    payment_response = {
        'razorpay_signature': params[:razorpay_signature],
    'order_id': params[:razorpay_order_id],
        'razorpay_payment_id': params[:razorpay_payment_id]
    }
    if Razorpay::Utility.verify_payment_signature(payment_response)
        #do something on success
    else
        #do something on failure    
    end
end
```

**Comment by jarednorman** [comment]
I'm not familiar with this particular gateway, but have you taken a look at the [spree-razorpay](https://github.com/neerajkumar/spree-razorpay) extension and they handle this? There are also other gateways like PayPal or Affirm that have similar flows, so depending on the specifics you might be able to use one of those as a reference for how this should work.

**Comment by rainerdema** [comment]
I'm not sure how the implementation was done and how this gateway works 🤔 
However, if you want to follow another solution for a solid integration with Solidus, you basically need a:
- Dedicated Payment Method
- Payment Source (with a dedicated table on the DB)
- Gateway

Right now, I see you are probably trying to do a gateway operation. This is usually done via [`active_merchant`](https://github.com/activemerchant/active_merchant/) in a dedicated payment method gateway.
With this flow, all the supported payment service provider gems have been basically developed:
https://guides.solidus.io/developers/payments/payment-service-providers.html

I noticed that for `Razorpay,` this gateway is not supported, but you can implement this if you want:
https://github.com/activemerchant/active_merchant/search?q=Razorpay&type=issues

To get an idea, you can take a look on:
- The current structure of the [basic credit card on Solidus](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/credit_card.rb)
- [solidus_paypal_commerce_platform](https://github.com/solidusio-contrib/solidus_paypal_commerce_platform/tree/master/app/models/solidus_paypal_commerce_platform)
- [solidus_stripe](https://github.com/solidusio/solidus_stripe/blob/master/app/models/spree/payment_method/stripe_credit_card.rb), where an `active_merchant gateway` is always used, but then it's managed into the payment method.
- [One of the other supported gems](https://guides.solidus.io/developers/payments/payment-service-providers.html)

When you have these components and the [dedicated partial views](https://github.com/solidusio/solidus_paypal_braintree/tree/master/app/views/spree), you will have the possibility to interact with the gateway even on the admin side, for example, for capture, refund, or void a payment.

**Comment by DanielePalombo** [comment]
Sorry to jump on it so late, but there is a solidus_extension for [razorpay](https://github.com/piyushswain/solidus_razorpay).
It's almost done, I think you can consider to use it.
